activerecord/test/cases/store_test.rb

```diff
@@ -191,4 +191,42 @@ class StoreTest < ActiveRecord::TestCase
     second_dump = YAML.dump(loaded)
     assert_equal @john, YAML.load(second_dump)
   end
+
+  test "setting type casted values on build" do
+    jamie = Admin::User.new(name: "Jamie", active: "true", salary: 3.1999, birthday: '2000-01-01')
+    assert_equal true, jamie.active
+    assert_equal 3, jamie.salary
+    assert_equal Date.new(2000,1,1), jamie.birthday
+  end
+
+  test "setting type casted values from db" do
+    jamie = Admin::User.new(name: "Jamie")
+    jamie.active = "t"
+    jamie.salary = 3.1999 
+    jamie.birthday = '01/01/2000'
+    jamie.save!
+
+    jamie = Admin::User.where(name: "Jamie").first
+   
+    assert_equal '2000-01-01', jamie.json_data['birthday']    
```

Comment:
I was thinking about the above solution and ended up with the following.

Assume that we've decorated our store with `TypedStore` (maybe several times, for each typed accessor). Now we can handle reading from and writing to db, and even updating entire store (_from user input_). 
But we still have to deal with writing thru accessors and store itself. Example:

```ruby
class User < AR
  typed_store_accessor :data, birthday: :date
end

user = User.create!(data: {birthday: '2015-02-23' )
user.reload
user.birthday.is_a?(Date) #=> true
user.data['birthday'].is_a?(Date) #=> true too

# But
user.birthday = '2015-02-25' 
user.birthday.class # => String
```
Because our `TypedStore` does nothing with the store hash itself, but accessor do.  
We want the value to be `Date` as with usual attribute.

Ok, we can do type casting within accessors (just like it is now). 
Then
```ruby
user.birthday = '2015-02-25' 
user.birthday.is_a?(Date) # => true
# But
user.data['birthday'] = '2015-02-26'
user.birthday.class  # => String again
```

Now we can say: "Never use direct store access if you have typed accessors". Never say never. 
This logic is very error-prone.

Store is just a Hash. "Maybe TypedStore should return smth like TypedHash?" No, It's not worth it.

**Conclusion**
There is no good (=not overcomplicated) solution to the problem.  

**P.S.**
All that I wanted to do it to replace [this example](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/store.rb#L57) with some short and generic method. Why to override accessors if we can put simple type casting inside of them (and don't care about store itself)?

cc @rafaelfranca @sgrif